Mathlib/MeasureTheory/SetSemiring.lean

```diff
@@ -0,0 +1,279 @@
+/-
+Copyright (c) 2023 Rémy Degenne. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Rémy Degenne, Peter Pfaffelhuber
+-/
+import Mathlib.Data.Set.Pairwise.Lattice
+import Mathlib.Data.Real.ENNReal
+import Mathlib.MeasureTheory.PiSystem
+
+/-! # Semirings of sets
+
+A semi-ring of sets `C` (in the sense of measure theory) is a family of sets containing `∅`,
+stable by intersection and such that for all `s, t ∈ C`, `t \ s` is equal to a disjoint union of
+finitely many sets in `C`. Note that a semi-ring of sets may not contain unions.
+
+An important example of a semi-ring of sets is intervals in `ℝ`. The intersection of two intervals
+is an interval (possibly empty). The union of two intervals may not be an interval.
+The set difference of two intervals may not be an interval, but it will be a disjoint union of
+two intervals.
+
+## Main definitions
+
+* `MeasureTheory.SetSemiring C`: property of being a semi-ring of sets.
+* `MeasureTheory.SetSemiring.diffFinset hs ht`: for `s, t` in a semi-ring `C`
+  (with `hC : SetSemiring C`) with `hs : s ∈ C`, `ht : t ∈ C`, this is a `Finset` of
+  pairwise disjoint sets such that `s \ t = ⋃₀ hC.diffFinset hs ht`.
+* `MeasureTheory.SetSemiring.diffFinset₀ hs hI`: for `hs : s ∈ C` and a finset `I` of sets in `C`
+  (with `hI : ↑I ⊆ C`), this is a `Finset` of pairwise disjoint sets such that
+  `s \ ⋃₀ I = ⋃₀ hC.diffFinset₀ hs hI`.
+
+## Main statements
+
+* `MeasureTheory.SetSemiring.exists_disjoint_finset_diff_eq`: the existence of the `Finset` given
+  by the definition `SetSemiring.diffFinset₀` (see above).
+
+-/
+
+open Finset Set
+
+open scoped ENNReal BigOperators
+
+namespace MeasureTheory
+
+variable {α : Type*} {C : Set (Set α)} {s t : Set α}
+
+/-- A semi-ring of sets `C` is a family of sets containing `∅`, stable by intersection and such that
+for all `s, t ∈ C`, `t \ s` is equal to a disjoint union of finitely many sets in `C`. -/
+structure SetSemiring (C : Set (Set α)) : Prop where
```

Comment:
Perhaps
```suggestion
structure IsSetSemiring (C : Set (Set α)) : Prop where
```
since this is a `Prop`, and that reduces ambiguity with `_root_.SetSemiring`